proc.c

```diff
@@ -80,62 +80,34 @@ CLONESETUP(VALUE clone, VALUE obj)
 }
 
 static void
-block_mark(const struct rb_block *block)
-{
-    switch (vm_block_type(block)) {
-      case block_type_iseq:
-      case block_type_ifunc:
-        {
-            const struct rb_captured_block *captured = &block->as.captured;
-            RUBY_MARK_MOVABLE_UNLESS_NULL(captured->self);
-            RUBY_MARK_MOVABLE_UNLESS_NULL((VALUE)captured->code.val);
-            if (captured->ep && !UNDEF_P(captured->ep[VM_ENV_DATA_INDEX_ENV]) /* cfunc_proc_t */) {
-                rb_gc_mark(VM_ENV_ENVVAL(captured->ep));
-            }
-        }
-        break;
-      case block_type_symbol:
-        RUBY_MARK_MOVABLE_UNLESS_NULL(block->as.symbol);
-        break;
-      case block_type_proc:
-        RUBY_MARK_MOVABLE_UNLESS_NULL(block->as.proc);
-        break;
-    }
-}
-
-static void
-block_compact(struct rb_block *block)
+block_mark_and_move(struct rb_block *block)
 {
     switch (block->type) {
       case block_type_iseq:
       case block_type_ifunc:
         {
             struct rb_captured_block *captured = &block->as.captured;
-            captured->self = rb_gc_location(captured->self);
-            captured->code.val = rb_gc_location(captured->code.val);
+            rb_gc_mark_and_move(&captured->self);
+            rb_gc_mark_and_move(&captured->code.val);
+            if (captured->ep) {
+                rb_gc_mark_and_move((VALUE *)&captured->ep[VM_ENV_DATA_INDEX_ENV]);
+            }
         }
         break;
       case block_type_symbol:
-        block->as.symbol = rb_gc_location(block->as.symbol);
+        rb_gc_mark_and_move(&block->as.symbol);
         break;
       case block_type_proc:
-        block->as.proc = rb_gc_location(block->as.proc);
+        rb_gc_mark_and_move(&block->as.proc);
         break;
     }
 }
 
 static void
-proc_compact(void *ptr)
+proc_mark_and_move(void *ptr)
 {
     rb_proc_t *proc = ptr;
-    block_compact((struct rb_block *)&proc->block);
-}
-
-static void
-proc_mark(void *ptr)
-{
-    rb_proc_t *proc = ptr;
-    block_mark(&proc->block);
+    block_mark_and_move((struct rb_block *)&proc->block);
     RUBY_MARK_LEAVE("proc");
```

Comment:
It is strange that there is `RUBY_MARK_LEAVE("proc")` but no `RB_MARK_ENTER("proc")`.

Reply:
Good point, I'll just remove these since they're not too useful.